Update: Improve checkbox clickable area and styling

enejb opened the pull request:
Wraps the checkbox input in an additional span to expand the clickable area, improving accessibility and usability. Adds new styles for the expanded area, including hover effects and proper positioning.

This PR expands increase the click area of the checkbox and adds a hover backgound so that it is more clear to the user what checkbox will be selected.

This PR is a follow up to https://github.com/WordPress/gutenberg/pull/73146 


## What?
* Adds a span that is clickable so has a hover class that increases the click area of checkboxes.

## Why?
Checkboxes can be hard to click since the click area is so small and in Dataview this can be an issue since missclicking a checkbox selection can cause the user to loose the selection. 

## How?
We add a new span that wraps the current checkbox. We absolute position this new chechbox area so that it shows up in the same place as before. 

## Testing Instructions
Visit the storybook and noctice that eveything still works as before. 
But now when you have a checkbox you also see the new cirlcle background when you hover over a checkbox.

### Testing Instructions for Keyboard
Key board should work as before. 


## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/944e25b2-182d-4216-ab13-8614d8dad3e9

enejb committed:
Update: Improve checkbox clickable area and styling

Wraps the checkbox input in an additional span to expand the clickable area, improving accessibility and usability. Adds new styles for the expanded area, including hover effects and proper positioning.

jameskoster commented on packages/components/src/checkbox-control/style.scss:
```diff
@@ -73,6 +73,23 @@
+	inset: 0;
+	padding: 8px;
+	margin-left: -8px;
+	margin-top: -8px;
```
I guess these could be based on `--checkbox-input-size`?

jameskoster commented on packages/components/src/checkbox-control/style.scss:
```diff
@@ -73,6 +73,23 @@
+	margin-top: -8px;
+	width: calc(var(--checkbox-input-size) + 16px);
+	height: calc(var(--checkbox-input-size) + 16px);
+	transition: cubic-bezier(0.075, 0.82, 0.165, 1) background-color 150ms;
```
Is this based on existing animation profiles? Do you think it's necessary to animate?

aduth replied:
We should also consider enabling this styling within a `@media not (prefers-reduced-motion)` media query, since it doesn't seem especially essential.

Related: https://www.w3.org/WAI/WCAG21/Understanding/animation-from-interactions.html

fcoveram replied:
I don't think we should animate at this level. I would approach motion from higher UI levels, like content areas, rather than from components. The former needs to convey more complex interactions and concepts, so when in pages where transitions exist, this component-level motion can feel overwhelming.

jameskoster commented on packages/components/src/checkbox-control/style.scss:
```diff
@@ -73,6 +73,23 @@
+.components-checkbox-control__input-container-area:hover {
+	cursor: pointer;
+	background-color: color-mix(in srgb, $components-color-accent 8%, $white);
+	border-radius: 50%;
```
I think there's a radius var you could use here.

jameskoster commented on packages/components/src/checkbox-control/style.scss:
```diff
@@ -73,6 +73,23 @@
+	padding: 8px;
+	margin-left: -8px;
+	margin-top: -8px;
+	width: calc(var(--checkbox-input-size) + 16px);
```
I think `+8px` would achieve the minimum required `24px` target size?

jameskoster commented:
I think it's kind of unavoidable that the checkbox will have a fairly unique treatment though because it's the only component that has a larger click target than the visible element. The only existing style I could think to mimic is the handle in range control. Maybe this would work:


<img width="205" height="132" alt="checkbox" src="https://github.com/user-attachments/assets/53396d22-2a56-45e9-b49f-a45b4974ea77" />

jasmussen commented:
Those all look okay enough to me, the point remains though, the color, shape, transition, opacity, all of those are values that would be good to systematize so that we have a catalog of materials to tap into. E.g. my point around radio and toggle feel valid here too: why just this one, and not all of them? And I'm still unconvinced we _need_ the hover style at all: the win here is the bigger click target, and that is indicated by the hand cursor. Similarly you can click the label next to the checkbox, that doesn't have a hover style despite toggling it. And finally, what does this look like in DataViews, where the entire row has a hover style, does it conflict with it or do they blend in?

All of those are solvable problems, and although it may read as a strong opinion, I assure you it's not. While I may have a preference going in one direction, if we feel confident we the code is clean and easy to update if we _do_ go back and systematize hover states at a later time, then we can move forward.

aduth commented on packages/components/src/checkbox-control/index.tsx:
```diff
@@ -110,37 +110,48 @@ export function CheckboxControl(
-						aria-describedby={ !! help ? id + '__help' : undefined }
+					<span
+						className="components-checkbox-control__input-container-area"
+						aria-hidden="true"
```
I haven't tested this yet, but my instinct is that `aria-hidden` will make this entire subtree (including the input) hidden to assistive technology.

enejb replied:
This is not applicable any more.

aduth commented on packages/components/src/checkbox-control/index.tsx:
```diff
@@ -110,37 +110,48 @@ export function CheckboxControl(
+					<span
+						className="components-checkbox-control__input-container-area"
+						aria-hidden="true"
 						onClick={ ( event ) => {
```
I would love if we could find a way to do this more natively without an extra `span` wrapper and JavaScript event handlers, though I think without being able to wrap `label`, our options are pretty limited. There might be some arcane CSS wizardry that would allow us to do it, but it seems like this might be the next best option.

That being said, a few suggestions on the implementation here:

- Rather than `firstElementChild`, can we rely on the `ref` assignment from the `<input />` element
   - Alternatively, rather than redirecting the click event to the `<input />`, we could duplicate the logic, i.e. call `onClick` and focus (via `ref`) the input
- Do we need the `stopPropagation` ?
- Is it possible to flatten this into the parent `<span />` so we don't need 2 separate elements?

enejb replied:
Thanks for suggestion a more CSS only approach here which seems to have worked well

fcoveram commented:
For the purpose of this PR, I think increasing the target size sounds the safer path. But I'm curious about the rationale for the behavior in DataViews and DataViewsPicker. I don't know the context of how it began.

enejb committed:
Css only try

enejb committed:
revert file

enejb committed:
More refinement

enejb commented:
Thanks for all the reviews on this PR. 

I removed the animation and the new span element.

I made it work with CSS only changes. The checkboxes did not change in size but they are the surface area of the checkbox is larger. 

I added the gray background here to show what it looks like. 

The regular PR doesn't add such background. 

Mobile:
<img width="296" height="121" alt="Screenshot 2025-11-27 at 3 08 20 PM" src="https://github.com/user-attachments/assets/c14a4470-2f9a-4e27-a77b-9f9c470f3030" />

Desktop:
<img width="294" height="94" alt="Screenshot 2025-11-27 at 3 08 16 PM" src="https://github.com/user-attachments/assets/1a2dc0df-07e2-461d-8f5e-8c8253236cda" />

It also means that the area between the label and the checkbox is clickable before this was not the case.

jasmussen commented:
> I added the gray background here to show what it looks like.
> The regular PR doesn't add such background.

Nice, wfm.

jameskoster commented:
It still seems a bit strange (imo) to solve one issue by introducing another, especially as hover styles are easy to change. But I agree it's not a huge problem, so I'm happy to merge revisit if that's the preferred approach.

enejb commented:
How do we want to proceed with this? Is the design review done? 

Do we need anther code review? cc @mirka and @aduth

mirka commented:
I do still think we're trying to fix a problem at the wrong layer.

<img width="109" height="82" alt="checkbox list" src="https://github.com/user-attachments/assets/e1ab65a3-7b7f-437e-80e4-4d37b032cb7f" />

For example, let's take this standard list of checkbox items. If we extend the click target of each checkbox 8px in each direction, those click targets will actually be overlapping each other. That doesn't make sense.

If we increase the click target like this, we'll also need to increase the clearance between each checkbox to at least 20px, maybe more. Is it really worth this density trade-off, given that this doesn't even really "fix" the original problem with the DataViews?

Indeed, [MUI](https://mui.com/material-ui/react-checkbox/) is showing a very large clearance between checkbox items, and a zone of ambiguity in the middle 😄

https://github.com/user-attachments/assets/58c91f68-9e33-438f-b923-79abc6e4b058
